File: microon-services/microon-services-repository/src/main/scala/microon/services/repository/api/scala/RepositoryService.scala
package microon.services.repository.api.scala

import java.io.Serializable
import java.util.concurrent.Future
import org.springframework.data.domain.{Pageable, Page, Sort}
import microon.spi.scala.activeobject.Void

trait RepositoryService[T, ID <: Serializable] {

  def save[S <: T](entity: S): Future[S]

  def save[S <: T](entities: Seq[S]): Future[Seq[S]]

  def findOne(id: ID): Future[T]

  def exists(id: ID): Future[Boolean]

  def findAll: Future[Seq[T]]

  def findAll(ids: Seq[ID]): Future[Seq[T]]

  def count: Future[Long]

  def delete(id: ID): Future[Void]

  def delete(entity: T): Future[Void]

  def delete(entities: Seq[_ <: T]): Future[Void]

  def deleteAll(): Future[Void]

  def findAll(sort: Sort): Future[Seq[T]]

  def findAll(pageable: Pageable): Future[Page[T]]

  def countByQuery(query: Any): Future[Long]

  def findAllByQuery(query: Any): Future[Seq[T]]

  def findAllByQuery(query: Any, pageable: Pageable): Future[Page[T]]

  def findAllByQuery(query: Any, sort: Sort): Future[Seq[T]]

  def findOneByQuery(query: Any): Future[T]

}

File: microon-services/microon-services-repository/src/main/java/microon/services/repository/api/java/RepositoryService.java
package microon.services.repository.api.java;

import microon.spi.scala.activeobject.Void;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;

import java.io.Serializable;
import java.util.concurrent.Future;

public interface RepositoryService<T, ID extends Serializable> {

    <S extends T> Future<S> save(S entity);

    <S extends T> Future<Iterable<S>> saveMany(Iterable<S> entities);

    Future<T> findOne(ID id);

    Future<Boolean> exists(ID id);

    Future<Iterable<T>> findAll();

    Future<Iterable<T>> findAll(Iterable<ID> ids);

    Future<Long> count();

    Future<Void> delete(ID id);

    Future<Void> delete(T entity);

    Future<Void> delete(Iterable<? extends T> entities);

    Future<Void> deleteAll();

    Future<Iterable<T>> findAll(Sort sort);

    Future<Page<T>> findAll(Pageable pageable);

    Future<Long> countByQuery(Object query);

    Future<Iterable<T>> findAllByQuery(Object query);

    Future<Page<T>> findAllByQuery(Object query, Pageable pageable);

    Future<Iterable<T>> findAllByQuery(Object query, Sort sort);

    Future<T> findOneByQuery(Object query);

}